File: 프로그래머스/연습문제/Lv2/야근 지수/야근 지수.py
from collections import Counter

def solution(n, works):
    if sum(works) <= n: return 0
    else:
        m = dict(Counter(works))
        while n > 0:
            m = dict(sorted(m.items(), key=lambda x:x[0], reverse=True))
            for k,v in m.items():
                if v <= n:
                    n -= v
                    if k-1 in m:
                        m[k-1] += v
                    else: m[k-1] = v
                    del m[k]
                else:
                    m[k] -= n
                    if k-1 in m:
                        m[k-1] += n
                    else: m[k-1] = n
                    n = 0

                break

        answer = 0
        for k,v in m.items():
            answer += (k**2)*v

        return answer

File: 프로그래머스/연습문제/Lv2/야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <map>
using namespace std;

long long solution(int n, vector<int> works) {
    map<int, int, greater<int>> m;    // <작업량, 개수>
    int sum = 0;
    
    for(int work : works){
        m[work]++;
        sum += work;
    }
    
    // 야근 안 해도 되면 0 리턴
    if(sum <= n) return 0;
    else{
        // 최대한 작업량 많은 것부터 일하기
        map<int, int>::iterator it;
        while(n){
            it = m.begin();
            if(it->second <= n){
                n -= it->second;
                m[it->first-1] += it->second;
                m.erase(it);
            }
            else{
                m[it->first] -= n;
                m[it->first-1] += n;
                n = 0;
            }
        }
    }
    
    // 남은 작업량 계산
    long long answer = 0;
    map<int, int>::iterator it;
    for(it=m.begin(); it!=m.end(); it++)
        answer += (it->first)*(it->first)*(it->second);
    
    return answer;
}
